src/backend/distributed/operations/shard_split.c

```diff
@@ -1145,3 +1221,650 @@ TryDropSplitShardsOnFailure(HTAB *mapOfShardToPlacementCreatedByWorkflow)
+																	  superUser,
+																	  databaseName);
+	ClaimConnectionExclusively(sourceConnection);
+
+	HTAB *mapOfShardToPlacementCreatedByWorkflow =
+		CreateEmptyMapForShardsCreatedByWorkflow();
+
+	/* Non-Blocking shard split workflow starts here */
+	PG_TRY();
+	{
+		/* 1) Physically create split children. */
+		CreateSplitShardsForShardGroup(mapOfShardToPlacementCreatedByWorkflow,
+									   shardGroupSplitIntervalListList,
+									   workersForPlacementList);
+
+		/*
+		 * 2) Create dummy shards due logical replication constraints.
+		 *    Refer to the comment section of 'CreateDummyShardsForShardGroup' for indepth
+		 *    information.
+		 */
+		CreateDummyShardsForShardGroup(
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			sourceShardToCopyNode,
+			workersForPlacementList);
+
+		/* 3) Create Publications. */
+		CreateShardSplitPublications(sourceConnection, shardSplitHashMapForPublication);
+
+		/*
+		 * 4) Create template replication Slot. It returns a snapshot. The snapshot remains
+		 * valid till the lifetime of the session that creates it. The connection is closed
+		 * at the end of the workflow.
+		 */
+		MultiConnection *templateSlotConnection = NULL;
+		char *snapShotName = CreateTemplateReplicationSlotAndReturnSnapshot(
+			shardIntervalToSplit, sourceShardToCopyNode, &templateSlotConnection);
+
+
+		/* 5) Do snapshotted Copy */
+		DoSplitCopy(sourceShardToCopyNode, sourceColocatedShardIntervalList,
+					shardGroupSplitIntervalListList, workersForPlacementList,
+					snapShotName);
+
+		/* 6) Execute 'worker_split_shard_replication_setup UDF */
+		List *replicationSlotInfoList = ExecuteSplitShardReplicationSetupUDF(
+			sourceShardToCopyNode,
+			sourceColocatedShardIntervalList,
+			shardGroupSplitIntervalListList,
+			workersForPlacementList);
+
+		/*
+		 * Subscriber flow starts from here.
+		 * Populate 'ShardSplitSubscriberMetadata' for subscription management.
+		 */
+		List *shardSplitSubscribersMetadataList =
+			PopulateShardSplitSubscriptionsMetadataList(
+				shardSplitHashMapForPublication, replicationSlotInfoList);
+
+		/* Create connections to the target nodes. TODO: can be refactored */
+		List *targetNodeConnectionList = CreateTargetNodeConnectionsForShardSplit(
+			shardSplitSubscribersMetadataList,
+			connectionFlags,
+			superUser, databaseName);
+
+		/* 7) Create copies of template replication slot */
+		char *templateSlotName = ShardSplitTemplateReplicationSlotName(
+			shardIntervalToSplit->shardId);
+		CreateReplicationSlots(sourceConnection, templateSlotName,
+							   shardSplitSubscribersMetadataList);
+
+		/* 8) Create subscriptions on target nodes */
+		CreateShardSplitSubscriptions(targetNodeConnectionList,
+									  shardSplitSubscribersMetadataList,
+									  sourceShardToCopyNode,
+									  superUser,
+									  databaseName);
+
+		/* 9) Wait for subscriptions to be ready */
+		WaitForShardSplitRelationSubscriptionsBecomeReady(
```

Comment:
Given we are not doing copy as part of subscription creation, this might not be necessary.

Reply:
After we create subscriptions, we ensure at this point that each of them start from the 'Ready' state. There are other states 
**State code: i = initialize, d = data is being copied, f = finished table copy, s = synchronized, r = ready (normal replication)**
We wont be having 'd' and 'f' but we would be transitioning from i->s->r. 

`WaitForShardSplitRelationSubscriptionsBecomeReady` will act as a barrier and ensures that subscriptions were ready. If something gets stuck in 'Initialize' state, we would not proceed and this would be logged thus helping in debugging.

Reply:
In my PR I indeed removed this function, because d and f were removed. But I think I agree that it makes sense to keep it.